Fix filename completion for pathnames that are relative to the user d…

heshrobe opened the pull request:
This address an issue that arises when trying to complete a pathname that is relative to the user's home directory, e.g. ~/foo.  This currently doesn't offer completions at all.  The submitted code, all in the single function "filename-completer" addresses and fixes the issue.  I've been using it for a while locally and don't see any misbehaviors.

dkochmanski commented on Core/clim-core/presentation-defs.lisp:
```diff
@@ -1607,77 +1607,110 @@ protocol retrieving gestures from a provided string."))
+	;; 	     (and
+	;; 	      (string-equal pn-name name)
+	;; 	      (let ((s (search type pn-type :test #'char-equal)))
+	;; 		(and s (zerop s)))))
```
please clean up this code before the review.

heshrobe commented:
I took care of this and updated the branch.

> On Apr 23, 2019, at 05:29, Daniel Kochmanski <notifications@github.com> wrote:
> 
> @dkochmanski requested changes on this pull request.
> 
> In Core/clim-core/presentation-defs.lisp:
> 
> > +	(setq completions (directory search-pathname)))
> +      ;; Now prune out all completions that don't start with the string
> +      ;; why is this necessary would complete-from-suggestions do this also?
> +      (let (;; (name (pathname-name actual-pathname))
> +       	    (type (pathname-type actual-pathname)))
> +	;; (loop for pn in completions
> +	;;    for pn-name = (pathname-name pn)
> +	;;    for pn-type = (pathname-type pn)
> +	;;    when (cond
> +	;; 	    ;; if the query has a type, then the name must be
> +	;; 	    ;; complete and match
> +	;; 	    (type
> +	;; 	     (and
> +	;; 	      (string-equal pn-name name)
> +	;; 	      (let ((s (search type pn-type :test #'char-equal)))
> +	;; 		(and s (zerop s)))))
> 
> please clean up this code before the review.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

Howard Shrobe

Principal Research Scientist and
Director CyberSecurity@CSAIL
MIT Computer Science and Artificial Intelligence Laboratory
hes@csail.mit.edu
617-253-7877 work
617-312-6439 cell

dkochmanski commented on Core/clim-core/presentation-defs.lisp:
```diff
@@ -1607,77 +1607,83 @@ protocol retrieving gestures from a provided string."))
+	(complete-from-possibilities original-string completions '(#\space)
+						    :action action
+						    :name-key #'namestring
+						    :value-key #'identity)))))
```
new method has a regression with completion (disregarding the home directory). example failure mode:

1. have only one file in the directory called `hello`
2. open listener and type ,Show File /path/to/hello/h
3. complete with tab

expected (and matching old behavior): completes to `hello`
observed (in the new code): completes to `H` (capital H)

heshrobe commented:
I’ve removed the third optional argument; it’s never invoked with it.


> On Apr 23, 2019, at 11:13, Daniel Kochmanski <notifications@github.com> wrote:
> 
> @dkochmanski requested changes on this pull request.
> 
> I'm still not done with testing a code, but I need to go now (and not sure when I'll get back to it), so will submit this partial review.
> 
> 
> +(defun filename-completer (string action &optional (default *default-pathname-defaults*))
> 
> function will never be called with the third optional argument, please remove it.
> 
> In Core/clim-core/presentation-defs.lisp:
> 
> 
> there is no need for double : accessor, this is climi package (remove package qualifier)

I fixed that

> '
> +			    :type (pathname-type pathname)
> +			    :directory new-directory)
> +	     )))
> 
> keep closing parenthesis in the same line.

Fixed that too.

> 
> 
> new method has a regression with completion (disregarding the home directory). example failure mode:
> 
> 	• have only one file in the directory called hello
> 	• open listener and type ,Show File /path/to/hello/h
> 	• complete with tab
> expected (and matching old behavior): completes to hello
> observed (in the new code): completes to H (capital H)
> 

I can’t replicate this.  In my machine (MacOS,  SBCL 1.5.1) or Unbuntu 18.04 with SBCL 1.5, it completes to hello.


> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

Howard Shrobe

Principal Research Scientist and
Director CyberSecurity@CSAIL
MIT Computer Science and Artificial Intelligence Laboratory
hes@csail.mit.edu
617-253-7877 work
617-312-6439 cell

heshrobe commented:
One more issue I just ran across: Different Lisps directory function treat  symlinks differently.  Allegro never resolves them.  SBCL provides a keyword argument :resolve-symlinks (which defaults to t, the opposite of Allegro).  I don’t have access to other Lisps, does anybody know what they do?

> On Apr 23, 2019, at 11:13, Daniel Kochmanski <notifications@github.com> wrote:
> 
> @dkochmanski requested changes on this pull request.
> 
> I'm still not done with testing a code, but I need to go now (and not sure when I'll get back to it), so will submit this partial review.
> 
> In Core/clim-core/presentation-defs.lisp:
> 
> > -             (t
> -              (values completed-string nil nil (length pathnames) nil))))
> -      (:complete
> -       ;; This is reached when input is activated, if we did
> -       ;; completion, that would mean that an input of "foo" would
> -       ;; be expanded to "foobar" if "foobar" exists, even if the
> -       ;; user actually *wants* the "foo" pathname (to create the
> -       ;; file, for example).
> -       (values so-far t so-far 1 nil))
> -      (:possibilities
> -       (values nil nil nil (length pathnames)
> -               (loop with length = (length directory-prefix)
> -                     for name in pathnames
> -                     collect (list (subseq (namestring name) length nil)
> -                                   name)))))))
> +(defun filename-completer (string action &optional (default *default-pathname-defaults*))
> 
> function will never be called with the third optional argument, please remove it.
> 
> In Core/clim-core/presentation-defs.lisp:
> 
> > -       (values so-far t so-far 1 nil))
> -      (:possibilities
> -       (values nil nil nil (length pathnames)
> -               (loop with length = (length directory-prefix)
> -                     for name in pathnames
> -                     collect (list (subseq (namestring name) length nil)
> -                                   name)))))))
> +(defun filename-completer (string action &optional (default *default-pathname-defaults*))
> +  (flet
> +      ((deal-with-home (pathname his-directory)
> +	 ;; SBCL (and maybe others) treat "~/xxx" specially, returning a pathname
> +	 ;; whose directory is (:ABSOLUTE :HOME xxx)
> +	 ;; But if you call Directory on that pathname the returned list
> +	 ;; are all complete pathnames without the :Home part!.
> +	 ;; So this replaces the :HOME with what it actually means
> +	 (let* ((home-env-variable (clim-internals::get-environment-variable "HOME"))
> 
> there is no need for double : accessor, this is climi package (remove package qualifier)
> 
> In Core/clim-core/presentation-defs.lisp:
> 
> > +	 ;; So this replaces the :HOME with what it actually means
> +	 (let* ((home-env-variable (clim-internals::get-environment-variable "HOME"))
> +		(home (loop for pos = 1 then (1+ next-pos)
> +			 for next-pos = (position #\/ home-env-variable :start pos)
> +			 collect (subseq home-env-variable pos next-pos)
> +			 until (null next-pos)))
> +		(new-directory (cons
> +				(first his-directory)
> +				(append home (rest (rest his-directory))))))
> +	     (make-pathname :host (pathname-host pathname)
> +			    :device (pathname-device pathname)
> +			    :name (pathname-name pathname)
> +			    :version (pathname-version pathname)
> +			    :type (pathname-type pathname)
> +			    :directory new-directory)
> +	     )))
> 
> keep closing parenthesis in the same line.
> 
> In Core/clim-core/presentation-defs.lisp:
> 
> > +					 :directory (butlast (pathname-directory pathname))
> +					 :name (first (last (pathname-directory pathname)))
> +					 :type nil)
> +			  new-completions))
> +		(t
> +		 (pushnew (make-pathname :host (pathname-host original-pathname)
> +					 :device (pathname-device original-pathname)
> +					 :directory (pathname-directory original-pathname)
> +					 :name (pathname-name pathname)
> +					 :type (pathname-type original-pathname))
> +			  new-completions))))
> +	    (setq completions (nreverse new-completions))))
> +	(complete-from-possibilities original-string completions '(#\space)
> +						    :action action
> +						    :name-key #'namestring
> +						    :value-key #'identity)))))
> 
> new method has a regression with completion (disregarding the home directory). example failure mode:
> 
> 	• have only one file in the directory called hello
> 	• open listener and type ,Show File /path/to/hello/h
> 	• complete with tab
> expected (and matching old behavior): completes to hello
> observed (in the new code): completes to H (capital H)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

Howard Shrobe

Principal Research Scientist and
Director CyberSecurity@CSAIL
MIT Computer Science and Artificial Intelligence Laboratory
hes@csail.mit.edu
617-253-7877 work
617-312-6439 cell

dkochmanski commented:
merged ,thanks! (and sorry for delay)

heshrobe commented:
Daniel, no problem.  Thanks for championing McClim!

> On May 28, 2019, at 04:23, Daniel Kochmanski <notifications@github.com> wrote:
> 
> merged ,thanks! (and sorry for delay)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

Howard Shrobe

Principal Research Scientist and
Director CyberSecurity@CSAIL
MIT Computer Science and Artificial Intelligence Laboratory
hes@csail.mit.edu
617-253-7877 work
617-312-6439 cell
